### ISSUE TITLE
Full support for tokens in transformers

### ISSUE THREAD

**Issue by erezsh** [lark-parser/lark#446]
1. Transformers should visit tokens by default (based on issue #389) 

2. Replace lexer callbacks interface with token visitation (provided using transformer)

**Comment by erezsh** [comment]
Also, maybe support builtin values as an alternative to functions

i.e.

```python
class MyTransformer(Transformer):
    TRUE = True
    FALSE = False
    NULL = None
```
But that's probably not worth the trouble?

**Comment by MegaIng** [comment]
> Replace lexer callbacks interface with token visitation (provided using transformer)

This clashes with your solution for issue #406.

> But that's probably not worth the trouble?

I think a simple function/tool would be enough:
```python
def constant(value): 
     return lambda *_: value
```

This can then be used like this:
```python
class MyTransformer(Transformer):
    TRUE = constant(True)
    FALSE = constant(False)
    NULL = constant(None)
```
This has the added benefit of not being specialcased for these values and/or having a time overhead, while still being more explicit.

**Comment by erezsh** [comment]
> This clashes with your solution for issue #406.

In what way?

> I think a simple function/tool would be enough

That's a nice idea

**Comment by MegaIng** [comment]
If you remove lexer callbacks, you can not change the type of tokens. (Unless I am misunderstanding the names or misinterpreted your plans and you don't want to remove the callbacks)

**Comment by erezsh** [comment]
I'm confused. `lexer_callbacks` are just a dict of TYPE -> FUNCTION

For example `{"TRUE": constant(True)}`

Isn't this exactly equivalent to collecting them into a dict from a transformer instance?

**Comment by MegaIng** [comment]
From my understanding, `lexer_callbacks` happen before parsing, Transformer afterwards. This means that you can change the behaviour of the parser (As seen in your solution to 406), which is no longer possible if `lexer_callbacks` are removed.

**Comment by erezsh** [comment]
I think I see the source of the confusion.

Transformers are usually applied after parsing, but Lark also accepts a transformer instance (using the `Lark(..., transformer=...)` option), that currently executes the rules *during* the parsing. It's therefor consistent to make sure it will process the tokens during parsing as well.

**Comment by bilderbuchi** [comment]
As a somewhat fresh user, I find it kind of confusing that the transformer behaves differently depending on which of two calling options is used. Btw, does this mean that it's possible to define/use two different transformers, one for use during parsing, one for after?

**Comment by erezsh** [comment]
It's unavoidable. The embedded transformer option exists for performance reasons, but it's limited in what it can do. The regular transformer is very flexible, but it has to run on an existing tree, which makes it less efficient, both in memory and performance.

Of course you can use two transformers. You can use even ten, but only one of them can run during the parse phase. For most use-cases, I don't think you would notice any different in behavior.

**Comment by bilderbuchi** [comment]
Ah ok. Then I'll have to read up on the differences, a performance boost could always be useful.

**Comment by erezsh** [comment]
Implemented in commit #1815bd7fbda932933fed604579e955a8c4d7c021
